source/dofs/dof_tools_sparsity.cc

```diff
@@ -759,20 +759,24 @@ namespace DoFTools
     {
       // implementation of the same function in namespace DoFTools for
       // non-hp DoFHandlers
-      template <typename DoFHandlerType,
+      template <int dim,
+                int spacedim,
```

Comment:
I have to admit that I don't quite understand what this change is about -- aren't you implementing a function here that is unrelated to the one you declare in the .h file, given that they have different template arguments and function argument types?

Reply:
After the changes in the first commit of this pull-req the MSVC-build failed. The reason was that in the hp-case MVSC couldn't deduce whether it should call the function taking the generic DoFHandlerType or the specialization for the hp::DoFHandler. Honestly, I don't understand why. It compiled both with gcc and clang without warnings. Adding the changes in the second commit was the only way I could come up with that would definitely make the MSVC-build pass.

Reply:
But it doesn't do what you want :-)

Think about how that file looks like once you just `#include` the .h file into the .cc file:
```
// from the .h file
namespace DoFTools {
  template <typename DoFHandlerType>
  void some_function (const DoFHandlerType &dh);
}

// now what you have in the .cc file:
namespace DoFTools {
  template <int dim>
  void some_function (const DoFHandler<dim> &dh)
  {
    ... your implementation...
  }
}
```
In essence, you end up with two overloaded functions that happen to share the same name. If a user calls the first of these two, you will find that there is implementation of that function to be found; conversely, a user can't call the second function because it has never been declared in a header file.

So maybe MSVC compiles this, but it doesn't do what you hope.